### PR TITLE
Change ReflectMapEntities to operate on components before insertion

### LINKED ISSUE #14465: Observers doesn't pair well with mapped entities
## Bevy version

v0.14.0

## What you did

I have a component that stores an entity. The component implements and reflects `MapEntities`, so on scene deserialization it automatically maps entities inside it.

## What went wrong

When I implement `on_add` hook on such component, it triggers before `MapEntities` which results in invalid entity access.
It happens because we insert components here:
https://github.com/bevyengine/bevy/blob/eabb58aa0466dadc5d485483efad9a90b4feb285/crates/bevy_scene/src/scene.rs#L117-L123
But entities mapped later:
https://github.com/bevyengine/bevy/blob/eabb58aa0466dadc5d485483efad9a90b4feb285/crates/bevy_scene/src/scene.rs#L128-L131

I think we should map entities before the insertion.

### PR DESCRIPTION
**Pull request by jrobsonchase**
Previous PR https://github.com/bevyengine/bevy/pull/14549 was closed in error and couldn't be reopened since I had updated the branch :crying_cat_face: 

# Objective

Fixes #14465

## Solution

`ReflectMapEntities` now works similarly to `MapEntities` in that it works on the reflected value itself rather than the component in the world after insertion. This makes it so that observers see the remapped entities on insertion rather than the entity IDs from the scene.

`ReflectMapEntities` now works for both components and resources, so we only need the one.

## Testing

* New unit test for `Observer`s + `DynamicScene`s
* New unit test for `Observer`s + `Scene`s
* Open to suggestions for other tests!

---

## Migration Guide

- Consumers of `ReflectMapEntities` will need to call `map_entities` on values prior to inserting them into the world.
- Implementors of `MapEntities` will need to remove the `mappings` method, which is no longer needed for `ReflectMapEntities` and has been removed from the trait.

### CODE REVIEW

**Comment by jrobsonchase** [comment]
I've rebased this on #15425 since it's gaining traction and I like its approach way better than what I had here previously.

**Comment by jrobsonchase** [comment]
Welp, this has now been un-rebased on #15425 after realizing the approach there wouldn't work for `HashSet`s. So now we're back to the original plan.

**Comment by jrobsonchase** [comment]
@alice-i-cecile This is ready for review once more!

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/reflect/map_entities.rs [hunk 1]
```diff
@@ -1,107 +1,37 @@
...
-
-impl<R: crate::system::Resource + MapEntities> FromType<R> for ReflectMapEntitiesResource {
-    fn from_type() -> Self {
-        ReflectMapEntitiesResource {
-            map_entities: |world, entity_mapper| {
-                if let Some(mut resource) = world.get_resource_mut::<R>() {
-                    resource.map_entities(entity_mapper);
-                }
+            map_entities: |reflected, mut mapper| {
+                let mut concrete = C::from_reflect(reflected).expect("mismatched reflected type");
```
Comment (Shatur) [inline crates/bevy_ecs/src/reflect/map_entities.rs:32]: Let's use the [recommended](https://doc.rust-lang.org/std/result/enum.Result.html#recommended-message-style) message style.
```suggestion
                let mut concrete = C::from_reflect(reflected).expect("reflected type should match");
```

**Review by Shatur — CHANGES_REQUESTED** [review]
Looks good!
Left a small suggestion.

**Review by Shatur — APPROVED** [review]

**Pull request merged**